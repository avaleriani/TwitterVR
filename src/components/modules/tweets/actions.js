import moment from 'moment';
import { v4 as uuid } from 'uuid';

const scope = '@@tweets';

export const CREATE_TWEET = `${scope}/CREATE_TWEET`;

export const createTweet = payload => ({
  type: CREATE_TWEET,
  payload: {
    id: uuid(), // unique id of tweet
    createdAt: moment(), // datetime object of tweet's creation
    userId: null, // user's id
    text: '', // content of the tweet
    replyToId: null, // if the tweet is a reply, id to the original tweet
    ...payload,
  },
});
