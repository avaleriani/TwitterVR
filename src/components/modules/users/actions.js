import { v4 as uuid } from 'uuid';
import * as fromUsers from './reducer';

const scope = '@@users';

export const RECIEVE_AUTH = `${scope}/RECIEVE_AUTH`;
export const SIGNIN = `${scope}/SIGNIN`;
export const SIGNUP = `${scope}/SIGNUP`;
export const LOGIN = `${scope}/LOGIN`;
export const LOGOUT = `${scope}/LOGOUT`;

export const recieveAuth = () => ({
  type: RECIEVE_AUTH,
});

export const signin = id => ({
  type: SIGNIN,
  payload: {
    id,
  },
});

export const signup = username => ({
  type: SIGNUP,
  payload: {
    id: uuid(),
    username,
  },
});

export const login = username => (dispatch, getState) => {
  const { users } = getState();

  dispatch({ type: LOGIN, username });

  const userId = fromUsers.getUserIdByUsername(users, username);

  if (userId) {
    return dispatch(signin(userId));
  }

  return dispatch(signup(username));
};

export const logout = () => ({
  type: LOGOUT,
});
