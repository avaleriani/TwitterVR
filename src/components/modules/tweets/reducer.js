import * as actions from './actions';
import { combineReducers } from 'redux';

export const initialState = {
  allIds: [],
  byId: {},
};

export function allIds(state = initialState.allIds, action) {
  switch (action.type) {
    case actions.CREATE_TWEET:
      return [...state, action.payload.id];
    default:
      return state;
  }
}

export function byId(state = initialState.byId, action) {
  switch (action.type) {
    case actions.CREATE_TWEET:
      return {
        ...state,
        [action.payload.id]: action.payload,
      };
    default:
      return state;
  }
}

export default combineReducers({
  allIds,
  byId,
});

// Selectors
// Probably better to use Reselect
export const getTweetById = (state, id) => state.byId[id];
export const getRepliesById = (state, id) =>
  getAllTweets(state).filter(tweet => tweet.replyToId === id);
export const getAllTweets = state =>
  state.allIds.map(id => getTweetById(state, id));
