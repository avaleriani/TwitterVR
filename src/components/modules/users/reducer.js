import * as actions from './actions';
import { combineReducers } from 'redux';

export const initialState = {
  active: null, // id of current user
  allIds: [],
  byId: {},
};

export function active(state = initialState.active, action) {
  switch (action.type) {
    case actions.SIGNIN:
    case actions.SIGNUP:
      return action.payload.id;
    case actions.LOGOUT:
      return null;
    default:
      return state;
  }
}

export function allIds(state = initialState.allIds, action) {
  switch (action.type) {
    case actions.SIGNUP:
      return [...state, action.payload.id];
    case actions.SIGNIN:
    case actions.LOGOUT:
      return state;
    default:
      return state;
  }
}

export function byId(state = initialState.byId, action) {
  switch (action.type) {
    case actions.SIGNUP:
      return {
        ...state,
        [action.payload.id]: action.payload,
      };
    case actions.SIGNIN:
    case actions.LOGOUT:
      return state;
    default:
      return state;
  }
}

export default combineReducers({
  active,
  allIds,
  byId,
});

// Selectors
// Probably better to use Reselect
export const isAuthenticated = state => !!state.active;
export const getUserById = (state, id) => state.byId[id];
export const getUserIdByUsername = (state, username) =>
  Object.keys(state.byId).find(id => state.byId[id].username === username);
