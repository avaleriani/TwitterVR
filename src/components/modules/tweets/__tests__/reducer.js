import reducer, {
  initialState,
  allIds,
  byId,
  getTweetById,
  getAllTweets,
  getRepliesById,
} from '../reducer';
import * as actions from '../actions';

jest.mock('uuid', () => ({
  v4: () => '<id>',
}));

jest.mock('moment', () => () => '<datetime>');

describe('tweets reducer', () => {
  it('returns initial state', () => {
    expect(reducer(undefined, {})).toEqual(initialState);
    expect(allIds(undefined, {})).toEqual(initialState.allIds);
    expect(byId(undefined, {})).toEqual(initialState.byId);
  });

  it('create new tweet', () => {
    const payload = { text: 'Test tweet' };
    const state = reducer(initialState, actions.createTweet(payload));
    expect(state).toMatchSnapshot();
  });
});

describe('tweets selectors', () => {
  it('gets tweets by id', () => {
    const user = { id: 123, username: 'user' };
    const state = {
      byId: {
        [user.id]: user,
      },
    };

    expect(getTweetById(state, user.id)).toEqual(user);
    expect(getTweetById(state, 321)).toBeUndefined();
  });

  it('gets all tweets', () => {
    const tweet1 = { id: 1, text: 'tweet 1' };
    const tweet2 = { id: 2, text: 'tweet 2' };
    const state = {
      allIds: [tweet1.id, tweet2.id],
      byId: {
        [tweet1.id]: tweet1,
        [tweet2.id]: tweet2,
      },
    };

    expect(getAllTweets(state)).toEqual([tweet1, tweet2]);
  });

  it('gets replies by id', () => {
    const tweet = { id: 1, text: 'tweet 1' };
    const reply = { id: 2, text: 'tweet 2', replyToId: 1 };
    const state = {
      allIds: [tweet.id, reply.id],
      byId: {
        [tweet.id]: tweet,
        [reply.id]: reply,
      },
    };

    expect(getRepliesById(state, tweet.id)).toEqual([reply]);
  });
});
