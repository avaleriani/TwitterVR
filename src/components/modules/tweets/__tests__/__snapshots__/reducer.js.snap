// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`tweets reducer create new tweet 1`] = `
Object {
  "allIds": Array [
    "<id>",
  ],
  "byId": Object {
    "<id>": Object {
      "createdAt": "<datetime>",
      "id": "<id>",
      "replyToId": null,
      "text": "Test tweet",
      "userId": null,
    },
  },
}
`;
